traitschema/schema.py

```diff
@@ -152,17 +152,34 @@ def to_hdf(self, filename, mode='w', compression=None,
         with h5py.File(filename, mode) as hfile:
             for name in self.class_visible_traits():
                 trait = self.trait(name)
-                # tt = trait.trait_type
 
                 # Workaround for saving arrays containing unicode. When the
                 # data type is unicode, each element is encoded as utf-8
                 # before being saved to hdf5
-                # FIXME: is there a better way of determining stringyness?
                 data = getattr(self, name)
-                if encode_string_arrays:
-                    if trait.array is True and str(data.dtype).find("<U") != -1:
+                if trait.array is True and encode_string_arrays:
+                    # Encode each element of an array containing unicode
+                    # elements
+                    if (isinstance(data, np.recarray) is False and
+                            data.dtype.char == 'U'):
                         data = [s.encode(encoding) for s in data]
 
+                    elif isinstance(data, np.recarray):
```

Comment:
You're checking if this is a recarray twice. Please refactor this to either nest the conditionals or store an intermediate variable.